nemo/collections/asr/data/audio_to_text_lhotse_prompted.py

```diff
@@ -0,0 +1,166 @@
+    """
+    This dataset is based on :class:`~nemo.collections.asr.data.audio_to_text_lhotse.LhotseSpeechToTextBpeDataset`.
+    It is a Lhotse-style dataset that converts a mini-batch of Cuts into tensors.
+    The main difference from ``LhotseSpeechToTextBpeDataset`` is that we introduce
+    a special prompt format for multitask encoder-decoder models.
+
+    To perform the prompt formatting, we accept a ``prompt_format_fn``.
+    It's expected to accept:
+    * a ``CutSet`` which it will internally iterate over for utterances, and
+    * a ``TokenizerWrapper`` object that will be internally used to tokenize the utterances
+
+    Tokenized utterances will be extended with special prompt tokens according to ``prompt_format_fn`` logic.
+    We support cuts with multiple supervision segments -- their tokenized texts will be concatenated before we add the prompt tokens.
+    This is useful, for example, in code-switched scenarios where each segment is spoken in a different language.
+    """
+
+    def __init__(
+        self, tokenizer: TokenizerSpec, prompt_format_fn: Callable[[CutSet, TokenizerWrapper], Sequence[Sequence[int]]]
+    ):
+        super().__init__()
+        self.tokenizer = TokenizerWrapper(tokenizer)
+        self.load_audio = AudioSamples(fault_tolerant=True)
+        self.padding_value = self.tokenizer._tokenizer.pad_id
+        self.prompt_format_fn = prompt_format_fn
+
+    def __getitem__(self, cuts: CutSet) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        audio, audio_lens, cuts = self.load_audio(cuts)
+
+        tokens = self.prompt_format_fn(cuts, self.tokenizer)
+        tokens = [torch.as_tensor(t) for t in tokens]
+        token_lens = torch.tensor([t.size(0) for t in tokens], dtype=torch.long)
+        tokens = collate_vectors(tokens, padding_value=self.padding_value)
+
+        return audio, audio_lens, tokens, token_lens
+
+
+# Mapping from a string name to a known prompt formatter function.
+PROMPT_FORMAT_FNS = {}
+
+
+def registered_prompt_format_fn(prompt_fn: Callable[[CutSet, TokenizerWrapper], Sequence[Sequence[int]]]):
+    """
+    Decorator for registering prompt functions under a name.
+
+    Example::
+
+        >>> @registered_prompt_format_fn
+        ... def my_prompt(cuts, tokenizer):
+        ...     pass
+        ...
+        ... prompt_fn = get_prompt_format_fn("my_prompt")
+    """
+    global PROMPT_FORMAT_FNS
+
+    PROMPT_FORMAT_FNS[prompt_fn.__name__] = prompt_fn
+    return prompt_fn
+
+
+def get_prompt_format_fn(name: str) -> Callable[[CutSet, TokenizerWrapper], Sequence[Sequence[int]]]:
+    if name not in PROMPT_FORMAT_FNS:
+        raise ValueError(
+            f"Unknown prompt format function name: {name} " f"(must be one of: {list(PROMPT_FORMAT_FNS.keys())}"
+        )
+    return PROMPT_FORMAT_FNS[name]
+
+
+@registered_prompt_format_fn
+def canary(cuts: CutSet, tokenizer: TokenizerWrapper) -> Sequence[Sequence[int]]:
+    """
+
```

Comment:
We should improve the docs here to list canary special tokens and give some examples of fully formed prompts